### PR TITLE
BUGZ-868: Avoid SafeLanding signal-slot connection resource leak

### PR DESCRIPTION
**Pull request by AndrewMeadows**
https://highfidelity.atlassian.net/browse/BUGZ-868

This PR fixes a problem where the `SafeLanding` logic connects one signal to three identical slots on every domain connection... then never disconnects them.

### CODE REVIEW

File: interface/src/Application.cpp [hunk 1]
```diff
@@ -2498,6 +2498,10 @@ Application::Application(int& argc, char** argv, QElapsedTimer& startupTimer, bo
     pauseUntilLoginDetermined();
 }
 
+void Application::setFailedToConnectToEntityServer() {
+    _failedToConnectToEntityServer = true;
+}
```
Comment (AndrewMeadows) [inline interface/src/Application.cpp]: This is not new, but I moved its implementation from the h-file to cpp-file for easier prototyping/debugging.

File: interface/src/Application.cpp [hunk 2]
```diff
@@ -5911,7 +5915,7 @@ void Application::resetPhysicsReadyInformation() {
     _gpuTextureMemSizeStabilityCount = 0;
     _gpuTextureMemSizeAtLastCheck = 0;
     _physicsEnabled = false;
-    _octreeProcessor.startEntitySequence();
+    _octreeProcessor.startSafeLanding();
```
Comment (AndrewMeadows) [inline interface/src/Application.cpp]: Better names make for more readable code.

File: interface/src/Application.cpp [hunk 3]
```diff
@@ -6161,14 +6165,31 @@ void Application::updateSecondaryCameraViewFrustum() {
...
+        // process octree stats packets are sent in between full sends of a scene (this isn't currently true).
+        // We keep physics disabled until we've received a full scene and everything near the avatar in that
+        // scene is ready to compute its collision shape.
+        if (getMyAvatar()->isReadyForPhysics()) {
+            _physicsEnabled = true;
+            setIsInterstitialMode(false);
+            getMyAvatar()->updateMotionBehaviorFromMenu();
+        }
+    }
+}
```
Comment (AndrewMeadows) [inline interface/src/Application.cpp]: This was pulled out of `Application::update()` for readability and to enable correct logic flow.

File: interface/src/Application.cpp [hunk 4]
```diff
@@ -7143,13 +7156,17 @@ void Application::resettingDomain() {
...
+            _failedToConnectToEntityServer = false;
+            _octreeProcessor.stopSafeLanding();
+            _octreeProcessor.startSafeLanding();
+        } else if (_entityServerConnectionTimer.isActive()) {
             _entityServerConnectionTimer.stop();
-            _entityServerConnectionTimer.setInterval(ENTITY_SERVER_CONNECTION_TIMEOUT);
-            _entityServerConnectionTimer.start();
         }
+        _entityServerConnectionTimer.setInterval(ENTITY_SERVER_CONNECTION_TIMEOUT);
+        _entityServerConnectionTimer.start();
```
Comment (AndrewMeadows) [inline interface/src/Application.cpp]: More correct starting of `SafeLanding` when connecting to new entity-server.

File: interface/src/Application.h [hunk 5]
```diff
@@ -786,8 +787,6 @@ private slots:
     qint64 _gpuTextureMemSizeStabilityCount { 0 };
     qint64 _gpuTextureMemSizeAtLastCheck { 0 };
 
-    quint64 _lastPhysicsCheckTime { usecTimestampNow() }; // when did we last check to see if physics was ready
-
```
Comment (AndrewMeadows) [inline interface/src/Application.h]: This was cruft.

File: interface/src/octree/SafeLanding.cpp [hunk 6]
```diff
@@ -33,61 +33,45 @@ bool SafeLanding::SequenceLessThan::operator()(const int& a, const int& b) const
...
             Locker lock(_lock);
             _entityTreeRenderer = entityTreeRenderer;
             _trackedEntities.clear();
-            _trackingEntities = true;
             _maxTrackedEntityCount = 0;
+            _initialStart = INVALID_SEQUENCE;
+            _initialEnd = INVALID_SEQUENCE;
+            _sequenceNumbers.clear();
+            _trackingEntities = true;
+            _startTime = usecTimestampNow();
```
Comment (AndrewMeadows) [inline interface/src/octree/SafeLanding.cpp]: Prepare data member values **before** making connections below.

File: interface/src/octree/SafeLanding.cpp [hunk 7]
```diff
@@ -33,61 +33,45 @@ bool SafeLanding::SequenceLessThan::operator()(const int& a, const int& b) const
...
-void SafeLanding::stopEntitySequence() {
-    Locker lock(_lock);
-    _trackingEntities = false;
-    _maxTrackedEntityCount = 0;
-    _trackedEntityStabilityCount = 0;
-    _initialStart = INVALID_SEQUENCE;
-    _initialEnd = INVALID_SEQUENCE;
-    _trackedEntities.clear();
-    _sequenceNumbers.clear();
-}
```
Comment (AndrewMeadows) [inline interface/src/octree/SafeLanding.cpp]: This method wasn't deleted per-se.  It was renamed and further down.

File: interface/src/octree/SafeLanding.cpp [hunk 8]
```diff
@@ -97,32 +81,94 @@ void SafeLanding::deleteTrackedEntity(const EntityItemID& entityID) {
...
     }
 }
 
-bool SafeLanding::isLoadSequenceComplete() {
-    if ((isEntityLoadingComplete() && isSequenceNumbersComplete()) || qApp->failedToConnectToEntityServer()) {
+void SafeLanding::updateTracking() {
+    if (!_trackingEntities || !_entityTreeRenderer) {
+        return;
+    }
+
```
Comment (AndrewMeadows) [inline interface/src/octree/SafeLanding.cpp]: Note the bracketed context below.  We must be careful not to try to `_lock` twice, else `SafeLanding` will deadlock.

File: interface/src/octree/SafeLanding.cpp [hunk 9]
```diff
@@ -33,61 +33,45 @@ bool SafeLanding::SequenceLessThan::operator()(const int& a, const int& b) const
     return lessThanWraparound<OCTREE_PACKET_SEQUENCE>(a, b);
 }
 
-void SafeLanding::startEntitySequence(QSharedPointer<EntityTreeRenderer> entityTreeRenderer) {
-
+void SafeLanding::startTracking(QSharedPointer<EntityTreeRenderer> entityTreeRenderer) {
     if (!entityTreeRenderer.isNull()) {
         auto entityTree = entityTreeRenderer->getTree();
-        if (entityTree) {
+        if (entityTree && !_trackingEntities) {
```
Comment (AndrewMeadows) [inline interface/src/octree/SafeLanding.cpp]: See this added check for `!_trackingEntities`?  It is important.  It prevents the `connect()`'s below from being called multiple times when arriving in a domain.

File: interface/src/octree/SafeLanding.cpp [hunk 10]
```diff
@@ -97,32 +81,94 @@ void SafeLanding::deleteTrackedEntity(const EntityItemID& entityID) {
...
+    Locker lock(_lock);
+    _trackingEntities = false;
+    if (_entityTreeRenderer) {
+        auto entityTree = _entityTreeRenderer->getTree();
+        disconnect(std::const_pointer_cast<EntityTree>(entityTree).get(),
+            &EntityTree::addingEntity, this, &SafeLanding::addTrackedEntity);
+        disconnect(std::const_pointer_cast<EntityTree>(entityTree).get(),
+            &EntityTree::deletingEntity, this, &SafeLanding::deleteTrackedEntity);
+        _entityTreeRenderer.reset();
     }
```
Comment (AndrewMeadows) [inline interface/src/octree/SafeLanding.cpp]: Above is where we tear down any unnecessary `SafeLanding` connections once `MyAvatar` has landed safely.  This reduces the number of useless slot calls going forward.

File: interface/src/octree/SafeLanding.cpp [hunk 11]
```diff
@@ -97,32 +81,94 @@ void SafeLanding::deleteTrackedEntity(const EntityItemID& entityID) {
...
+                    (startIter != _sequenceNumbers.end() &&
+                     endIter != _sequenceNumbers.end() &&
+                     ((distance(startIter, endIter) == sequenceSize - 1) || !missingSequenceNumbers)));
+            }
+            if (shouldStop) {
+                stopTracking();
+            }
+        }
+    }
+}
```
Comment (AndrewMeadows) [inline interface/src/octree/SafeLanding.cpp]: This code in `SafeLanding::updateTracking()` above isn't new so much as moved out of two poorly named methods and combined into one `void` method to provide a more readable `SafeLanding` API.

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/20170/HighFidelity-Beta-PR15846-bf068a3.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/20170/HighFidelity-Beta-Interface-PR15846-bf068a3.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/20170/HighFidelity-Beta-PR15846-bf068a3.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/20170/HighFidelity-Beta-Interface-PR15846-bf068a3.exe

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20170/HighFidelity-Beta-PR15846-bf068a3.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20170/HighFidelity-Quest-Beta-PR15846-bf068a3.apk)

**Comment by AndrewMeadows** [comment]
This is not ready.  I'm testing domain transitions on this right now and think I have spotted a problem.

**Comment by AndrewMeadows** [comment]
My concern about changing domains was a false alarm.  This PR is ready to go.

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20174/HighFidelity-Beta-PR15846-0bec2bc.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20174/HighFidelity-Quest-Beta-PR15846-0bec2bc.apk)

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/20174/HighFidelity-Beta-PR15846-0bec2bc.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/20174/HighFidelity-Beta-Interface-PR15846-0bec2bc.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/20174/HighFidelity-Beta-PR15846-0bec2bc.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/20174/HighFidelity-Beta-Interface-PR15846-0bec2bc.exe

**Pull request merged**